Javas

Kimjinho-k opened the pull request:
2주차 과제 JavaScript를 이용한 가격 별 상품 추천 프로그램입니다.
git bash 사용 능력 부족으로 인해 상세하고 세밀한 commit을 하지 못한 점 양해 부탁 드립니다!..

Kimjinho-k committed:
feat: create html for JavaScript

Kimjinho-k committed:
feat: create main JavaScript

해당 코드 안에 주석 처리로 상세한 설명을 해놓았습니다.

heejinlee07 commented on JS/main.js:
```diff
@@ -0,0 +1,46 @@
+
+// 사용자 입력 받기
+const line = prompt("최대 금액을 입력해주세요.");
+const amount = +line;
```
prompt()의 결과가 문자열이라서 숫자로 변환하기 위해 `+`기호를 작성해주신 것 같습니다. 
일반적으로 코드 상에서 문자열을 숫자로 변환할 때는 `+`기호를 이용해서 강제 변환하는 것보다는 
javaScript 내장 함수를 이용하는 것이 좋습니다. 실무에서도 보통 아래 함수를 이용해서 
문자열을 숫자로 변환합니다.

- parseInt(line);
- Number(line);

보다 자세한 내용은 아래 링크를 참조하세요:)
https://www.delftstack.com/ko/howto/javascript/convert-string-to-number-javascript/

heejinlee07 commented on JS/main.js:
```diff
@@ -0,0 +1,46 @@
+const getItemByAmount = (data, amount) => {
+  let max = 0; // max를 0으로 초기화 -> msg의 삼항연산자 조건식 때문
+  data.map((found) => {
+    //반복문 map을 활용하여 data 배열의 모든 객체를 순회 하고 새로운 배열을 return
```
map은 반복문이라기보다 고차배열 메서드라고 보통 불립니다.

heejinlee07 commented on JS/main.js:
```diff
@@ -0,0 +1,46 @@
+  let max = 0; // max를 0으로 초기화 -> msg의 삼항연산자 조건식 때문
+  data.map((found) => {
+    //반복문 map을 활용하여 data 배열의 모든 객체를 순회 하고 새로운 배열을 return
+    if (amount > found.price) { // 입력 한 금액과 배열의 첫번 째 객체를 비교
+      if (max === 0) { // 위의 조건문이 flase가 되면 안되기 때문에 max를 found 리스트에 대입
+        max = found;  // max 값에 입력한 값과 비교 된 found.price 객체를 할당
+      } else if (max.price < found.price) { // max 값이 0이 아니라면 실행
+        max = found;
+      }
+    } else if (amount === found.price) { // 첫번째 if문의 조건이 false 일 경우 실행 
+      max = found;
+    }
+  });
+  return max;
+};
```
map을 이용하신 점은 아주 좋습니다. 하지만 filter를 이용하면 이 조건을 보다 간결하게 작성하실 수 있습니다. 
예를 들어 아래와 같이 작성한다면 data 배열의 price중 함수에서 입력받은 amount보다 작거나 같은 값만 남길 수 있습니다. 말 그대로 필터링할 수 있습니다. 
```
data.filter((element) => element.price <= amount);
```

heejinlee07 replied:
그리고 map, filter와 같은 고차배열함수는 현업에서 정말 많이 쓰는 매서드입니다. 아래 링크를 참고하여 공부해두시면 많이 도움이 되실 거예요. 단 push, pop 등 원본 배열을 손상시키는 배열은 가능하면 쓰지않아야 합니다!

https://poiemaweb.com/js-array-higher-order-function
https://react.vlpt.us/basic/13-array-insert.html

heejinlee07 reviewed (commented):
LGTM😃
고생많으셨습니다!

전반적으로 로직을 잘 구성해주셨는데요.
map 외에도 다른 고차배열 함수 사용법에 대해 익히시면
도움이 많이 될거예요.
고생하셨습니다!

Kimjinho-k committed:
fix: change the code

17번째 줄에 const amount = +line의 강제 변환 +line을javaScript 내장
함수인 parselnt(line)으로 변경하였습니다.

Kimjinho-k committed:
fix: change annotation

map 함수를 반복문이라고 주석 달았던 것을 고차배열 메소드로 변경
하였습니다.
